src/bootstrap/replication/peer_discovery.rs

```diff
@@ -0,0 +1,161 @@
+                        libp2p::gossipsub::Event::Message { message, .. }
+                    )) = event {
+                        if let Ok(gossip_msg) = proto::GossipMessage::decode(message.data.as_slice()) {
+                            if let Some(proto::gossip_message::GossipMessage::ContactInfoMessage(ci)) = gossip_msg.gossip_message {
+                                if let Some(body) = ci.body { self.handle_contact_info(body); }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn gossip_addr_to_http_addr(&self, gossip_addr: &str) -> Option<String> {
+        let parts: Vec<&str> = gossip_addr.split('/').collect();
+        if parts.len() >= 3 && parts[1] == "ip4" {
+            let ip = parts[2];
+            // When we have just the IP address, we'll assume HTTP and the default port
+            Some(format!("http://{}:{}", ip, DEFAULT_RPC_PORT))
+        } else {
+            None
+        }
+    }
+
+    fn handle_contact_info(&self, contact_info: ContactInfoBody) {
+        let peer_id = match PeerId::from_bytes(&contact_info.peer_id) {
+            Ok(p) => p,
+            Err(_) => {
+                warn!("Invalid peer id in contact info");
+                return;
+            }
+        };
+
+        let http_addr = if !contact_info.announce_rpc_address.is_empty() {
+            // This peer is announcing an RPC address, use that one
+            contact_info.announce_rpc_address
+        } else if !contact_info.gossip_address.is_empty() {
+            // Fallback to deriving from gossip address
+            match self.gossip_addr_to_http_addr(&contact_info.gossip_address) {
+                Some(addr) => addr,
+                None => {
+                    warn!(
+                        contact_info = %serde_json::to_string(&contact_info).unwrap_or_else(|_| "failed to serialize".to_string()),
+                        "Invalid gossip address format, not adding peer",
+                    );
+                    return;
+                }
+            }
+        } else {
+            warn!(
+                contact_info = %serde_json::to_string(&contact_info).unwrap_or_else(|_| "failed to serialize".to_string()),
+                "No RPC or gossip address provided in contact info, not adding peer"
+            );
+            return;
+        };
+
+        // Cooldown check
+        if let Some(ts) = self.incompatible_peers.lock().unwrap().get(&peer_id) {
+            if ts.elapsed() < INCOMPATIBLE_PEER_COOLDOWN {
+                return;
+            }
+        }
+
+        // Spawn validation (metadata height match) via RpcClientsManager::add_new_peer
+        let rpc_manager = self.rpc_manager.clone();
+        let incompatible = self.incompatible_peers.clone();
+        info!(peer = %peer_id, addr = %http_addr, "Discovered potential replication peer");
+        tokio::spawn(async move {
```

Comment:
Why do we need a thread for each of these calls?

Reply:
you're right - we need only one of the spawn. I've removed the one from add_peer
We need to keep this one because I didn't want to make handle_contact_info async and .await a network connection in the gossip messages handler. It is possible that the connection takes multiple seconds/stalls, and I didn't want to block the gossip selector in case the peer connection stalls.